zebra-consensus/src/block/subsidies.rs

```diff
@@ -0,0 +1,241 @@
+//! Calculation of Block Subsidy, Funding Streams, and Founders’ Reward
+
+use super::*;
+
+use zebra_chain::{
+    block::Height,
+    parameters::{Network, NetworkUpgrade::*},
+};
+
+use crate::parameters::{fs, Params};
+
+fn slow_start_shift() -> Height {
+    Height(Params::SLOW_START_INTERVAL / 2)
+}
+
+fn slow_start_rate() -> u32 {
+    Params::MAX_BLOCK_SUBSIDY / Params::SLOW_START_INTERVAL
+}
+
+fn is_bolossom_activated(height: block::Height, network: Network) -> Option<Height> {
+    let blossom_height = Blossom.activation_height(network)?;
+    let condition = height >= blossom_height;
+    match condition {
+        true => Some(blossom_height),
+        false => None,
+    }
+}
+
+fn halving(height: Height, network: Network) -> u32 {
+    let blossom_height = is_bolossom_activated(height, network);
+    let condition = height < slow_start_shift();
+    match condition {
+        true => 0,
+        false => match blossom_height {
+            Some(blossom_height) => {
+                let scaled_halvings = ((blossom_height.0 - slow_start_shift().0)
+                    * Params::BLOSSOM_POW_TARGET_SPACING_RATIO)
+                    + (height.0 - blossom_height.0);
+                scaled_halvings / Params::POST_BLOSSOM_HALVING_INTERVAL
+            }
+            _ => {
+                (((height.0 - slow_start_shift().0) / Params::PRE_BLOSSOM_HALVING_INTERVAL) as f32)
+                    .floor() as u32
+            }
+        },
+    }
+}
```

Comment:
translated to use the combinator functions on `Option`. I think this same treatment could be applied to most of the functions in this file but I'll leave updating the rest of them to you.

```suggestion
fn halving(height: Height, network: Network) -> u32 {
    if height < slow_start_shift() {
        return 0;
    }

    Blossom
        .activation_height(network)
        .filter(|&blossom_height| height >= blossom_height)
        .map(|blossom_height| {
            (blossom_height.0 - slow_start_shift().0) * Params::BLOSSOM_POW_TARGET_SPACING_RATIO
                + height.0
                - blossom_height.0
        })
        .map(|scaled_halvings| scaled_halvings / Params::POST_BLOSSOM_HALVING_INTERVAL)
        .unwrap_or_else(|| (height.0 - slow_start_shift().0) / Params::PRE_BLOSSOM_HALVING_INTERVAL)
}
```

Reply:
Refactored `halving` as suggested at https://github.com/ZcashFoundation/zebra/pull/1051/commits/1530c46193958a134a220f16a6b0ec7c4691c1bb. Other functions will be treated in separated commits.

Reply:
`block_subsidy` refactored similar to `halving` at https://github.com/ZcashFoundation/zebra/pull/1051/commits/5bc047d1b6f3d0b3ec63ad2fa791f9805d191f7f
In `founders_reward` and `funding_stream` functions i replaced the `match` patterns with just `if` and `else` as i think is the most clear thing to do there. Please correct me if i am wrong.